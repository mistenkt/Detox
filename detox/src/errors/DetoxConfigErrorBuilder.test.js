const _ = require('lodash');
const fs = require('fs');
const os = require('os');
const DetoxConfigErrorBuilder = require('./DetoxConfigErrorBuilder');

describe('DetoxConfigErrorBuilder', () => {
  /** @type DetoxConfigErrorBuilder */
  let builder;
  let build;

  beforeEach(() => {
    builder = new DetoxConfigErrorBuilder();
  });

  describe('.noConfigurationSpecified', () => {
    beforeEach(() => {
      build = () => builder.noConfigurationSpecified();
    });

    it('should create a generic error, if the configuration file is unknown', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create a generic error, if the configuration file is not a package.json', () => {
      builder.setDetoxConfigPath('detox.config.json');
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with a package.json hint, if THAT is the configuration file', () => {
      builder.setDetoxConfigPath('/home/detox/myproject/package.json');
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.noConfigurationAtGivenPath', () => {
    beforeEach(() => {
      build = () => builder.noConfigurationAtGivenPath();
      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
    });

    it('should create an error with the attempted config path', () => {
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.failedToReadConfiguration', () => {
    it('should create a generic error, if I/O error is unknown', () => {
      builder.setDetoxConfigPath('/etc/detox/config.js');
      expect(builder.failedToReadConfiguration()).toMatchSnapshot();
    });

    it('should create a simple error, but with the original intercepted IO error', () => {
      const ioError = _.attempt(() => fs.readFileSync(os.homedir()));
      builder.setDetoxConfigPath('/home/detox');
      expect(builder.failedToReadConfiguration(ioError)).toMatchSnapshot();
    });
  });

  describe('.noConfigurationsInside', () => {
    beforeEach(() => {
      build = () => builder.noConfigurationsInside();
    });

    it('should create a generic error if all is unknown', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with Detox config fragment, if the path is not known', () => {
      builder.setDetoxConfig({ config: { ios: {}, android: {} } });
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with Detox config location hint, if it is known', () => {
      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
      builder.setDetoxConfig({});

      expect(build()).toMatchSnapshot();
    });
  });

  describe('.cantChooseConfiguration', () => {
    beforeEach(() => {
      build = () => builder.cantChooseConfiguration();
      builder.setDetoxConfig({
        configurations: {
          conf1: {},
          conf2: {},
        },
      });
    });

    it('should create a generic error, if the config location is not known', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with a hint, if the config location is known', () => {
      builder.setDetoxConfigPath('/etc/detox/config.js');
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.noConfigurationWithGivenName', () => {
    beforeEach(() => {
      build = () => builder.noConfigurationWithGivenName();
      builder.setConfigurationName('otherConf')
      builder.setDetoxConfig({
        configurations: {
          conf1: {},
        },
      });
    });

    it('should create a generic error, if the config location is not known', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with a hint, if the config location is known', () => {
      builder.setDetoxConfigPath('/etc/detox/config.js');
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.configurationShouldNotBeEmpty', () => {
    beforeEach(() => {
      build = () => builder.configurationShouldNotBeEmpty();
      builder.setConfigurationName('empty')
      builder.setDetoxConfig({
        configurations: {
          nonEmpty: { launchArgs: { key: 'value' } },
          empty: {},
        },
      });
    });

    it('should create a generic error, if the config location is not known', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with a hint, if the config location is known', () => {
      builder.setDetoxConfigPath('/etc/detox/config.js');
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.thereAreNoDeviceConfigs', () => {
    beforeEach(() => {
      build = () => builder.thereAreNoDeviceConfigs('simulator');
      builder.setConfigurationName('conf1')
      builder.setDetoxConfig({
        configurations: {
          conf1: {
            device: 'simulator',
          },
        },
      });
    });

    it('should create an error with a hint', () => {
      expect(build()).toMatchSnapshot('without config path');

      builder.setDetoxConfigPath('/etc/detox/config.js');
      expect(build()).toMatchSnapshot('with config path');
    });
  });

  describe('.missingDeviceProperty', () => {
    beforeEach(() => {
      build = (alias) => builder.missingDeviceProperty(alias, ['foo', 'bar']);
      builder.setConfigurationName('android.release');
      builder.setDetoxConfig({
        devices: {
          'emulator': {
            type: 'android.emulator',
          },
        },
        configurations: {
          'android.release': {
            type: 'android.emulator',
          },
        },
      });
    });

    it('should work with plain configurations', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should work with aliased configurations', () => {
      delete builder.selectedConfiguration.type;
      builder.selectedConfiguration.device = 'emulator';

      expect(build('emulator')).toMatchSnapshot();
    });

    it('should include the config location into a hint message if it is known', () => {
      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.malformedAppLaunchArgs', () => {
    beforeEach(() => {
      build = () => builder.malformedAppLaunchArgs();
      builder.setConfigurationName('android.release');
      builder.setDetoxConfig({
        configurations: {
          'android.release': {
            type: 'android.emulator',
            utilBinaryPaths: '/valid/path/outside/of/array',
            device: 'Pixel 4',
            launchArgs: 'do not use strings here please',
          },
        },
      });
    });

    it('should create an error with specifying the config name', () => {
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.malformedUtilBinaryPaths', () => {
    beforeEach(() => {
      build = () => builder.malformedUtilBinaryPaths()
      builder.setConfigurationName('android.release');
      builder.setDetoxConfig({
        configurations: {
          'android.release': {
            type: 'android.emulator',
            utilBinaryPaths: '/valid/path/outside/of/array',
            device: 'Pixel 4',
          },
        },
      });
    });

    it('should create an error with specifying the config name', () => {
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.missingBuildScript', () => {
    beforeEach(() => {
      build = () => builder.missingBuildScript();
      builder.setConfigurationName('android.release');
      builder.setDetoxConfig({
        configurations: {
          'android.release': {
            type: 'android.emulator',
            device: {
              avdName: 'Pixel_2_API_29'
            }
          }
        }
      });
    });

    it('should create a generic error, if the config location is not known', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with a hint, if the config location is known', () => {
      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.invalidServerProperty', () => {
    beforeEach(() => {
      build = () => builder.invalidServerProperty();
      builder.setConfigurationName('android.release');
      builder.setDetoxConfig({
        session: {
          server: 'localhost',
        },
        configurations: {
          'android.release': {
            type: 'android.emulator',
            device: {
              avdName: 'Pixel_2_API_29'
            }
          }
        }
      });
    });

    it('should create a generic error, if the config location is not known', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with a hint, if the config location is known', () => {
      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.invalidSessionId', () => {
    beforeEach(() => {
      build = () => builder.invalidSessionIdProperty();
      builder.setConfigurationName('android.release');
      builder.setDetoxConfig({
        configurations: {
          'android.release': {
            type: 'android.emulator',
            device: {
              avdName: 'Pixel_2_API_29',
            },
            session: {
              sessionId: 234589798234,
            },
          }
        }
      });
    });

    it('should create a generic error, if the config location is not known', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with a hint, if the config location is known', () => {
      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
      expect(build()).toMatchSnapshot();
    });

    it('should point to global session if there is one', () => {
      builder.setDetoxConfig({
        session: {
          server: 'ws://localhost:12837',
        },
        configurations: {},
      })

      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
      expect(build()).toMatchSnapshot();
    });
  });

  describe('.invalidDebugSynchronizationProperty', () => {
    beforeEach(() => {
      build = () => builder.invalidDebugSynchronizationProperty();
      builder.setConfigurationName('android.release');
      builder.setDetoxConfig({
        configurations: {
          'android.release': {
            type: 'android.emulator',
            device: {
              avdName: 'Pixel_2_API_29',
            },
            session: {
              debugSynchronization: '3000',
            },
          }
        }
      });
    });

    it('should create a generic error, if the config location is not known', () => {
      expect(build()).toMatchSnapshot();
    });

    it('should create an error with a hint, if the config location is known', () => {
      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
      expect(build()).toMatchSnapshot();
    });

    it('should point to global session if there is one', () => {
      builder.setDetoxConfig({
        session: {
          server: 'ws://localhost:12837',
        },
        configurations: {},
      })

      builder.setDetoxConfigPath('/home/detox/myproject/.detoxrc.json');
      expect(build()).toMatchSnapshot();
    });
  });
});
