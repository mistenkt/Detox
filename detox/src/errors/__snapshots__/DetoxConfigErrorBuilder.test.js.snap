// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetoxConfigErrorBuilder .cantChooseConfiguration should create a generic error, if the config location is not known 1`] = `
[DetoxConfigError: Cannot determine which configuration to use from Detox config.

HINT: Use --configuration to choose one of the following:
* conf1
* conf2]
`;

exports[`DetoxConfigErrorBuilder .cantChooseConfiguration should create an error with a hint, if the config location is known 1`] = `
[DetoxConfigError: Cannot determine which configuration to use from Detox config at path:
/etc/detox/config.js

HINT: Use --configuration to choose one of the following:
* conf1
* conf2]
`;

exports[`DetoxConfigErrorBuilder .configurationShouldNotBeEmpty should create a generic error, if the config location is not known 1`] = `
[DetoxConfigError: Cannot use an empty configuration "empty".

HINT: A valid configuration should have "device" and "app" properties defined.
Examine your Detox config.

{
  configurations: {
    empty: {},
    nonEmpty: [Object]
  }
}]
`;

exports[`DetoxConfigErrorBuilder .configurationShouldNotBeEmpty should create an error with a hint, if the config location is known 1`] = `
[DetoxConfigError: Cannot use an empty configuration "empty".

HINT: A valid configuration should have "device" and "app" properties defined.
Examine your Detox config at path:
/etc/detox/config.js

{
  configurations: {
    empty: {},
    nonEmpty: [Object]
  }
}]
`;

exports[`DetoxConfigErrorBuilder .failedToReadConfiguration should create a generic error, if I/O error is unknown 1`] = `
[DetoxConfigError: An error occurred while trying to load Detox config from:
/etc/detox/config.js]
`;

exports[`DetoxConfigErrorBuilder .failedToReadConfiguration should create a simple error, but with the original intercepted IO error 1`] = `
[DetoxConfigError: An error occurred while trying to load Detox config from:
/home/detox

EISDIR: illegal operation on a directory, read]
`;

exports[`DetoxConfigErrorBuilder .invalidDebugSynchronizationProperty should create a generic error, if the config location is not known 1`] = `
[DetoxConfigError: session.debugSynchronization should be a positive number

HINT: Check that in your Detox config.

{
  configurations: {
    'android.release': {
      session: {
        debugSynchronization: '3000'
      }
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .invalidDebugSynchronizationProperty should create an error with a hint, if the config location is known 1`] = `
[DetoxConfigError: session.debugSynchronization should be a positive number

HINT: Check that in your Detox config at path:
/home/detox/myproject/.detoxrc.json

{
  configurations: {
    'android.release': {
      session: {
        debugSynchronization: '3000'
      }
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .invalidDebugSynchronizationProperty should point to global session if there is one 1`] = `
[DetoxConfigError: session.debugSynchronization should be a positive number

HINT: Check that in your Detox config at path:
/home/detox/myproject/.detoxrc.json

{
  session: {
    server: 'ws://localhost:12837'
  }
}]
`;

exports[`DetoxConfigErrorBuilder .invalidServerProperty should create a generic error, if the config location is not known 1`] = `
[DetoxConfigError: session.server property is not a valid WebSocket URL

HINT: Expected something like "ws://localhost:8099".
Check that in your Detox config.

{
  session: {
    server: 'localhost'
  },
  configurations: {
    'android.release': {}
  }
}]
`;

exports[`DetoxConfigErrorBuilder .invalidServerProperty should create an error with a hint, if the config location is known 1`] = `
[DetoxConfigError: session.server property is not a valid WebSocket URL

HINT: Expected something like "ws://localhost:8099".
Check that in your Detox config at path:
/home/detox/myproject/.detoxrc.json

{
  session: {
    server: 'localhost'
  },
  configurations: {
    'android.release': {}
  }
}]
`;

exports[`DetoxConfigErrorBuilder .invalidSessionId should create a generic error, if the config location is not known 1`] = `
[DetoxConfigError: session.sessionId property should be a non-empty string

HINT: Check that in your Detox config.

{
  configurations: {
    'android.release': {
      session: {
        sessionId: 234589798234
      }
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .invalidSessionId should create an error with a hint, if the config location is known 1`] = `
[DetoxConfigError: session.sessionId property should be a non-empty string

HINT: Check that in your Detox config at path:
/home/detox/myproject/.detoxrc.json

{
  configurations: {
    'android.release': {
      session: {
        sessionId: 234589798234
      }
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .invalidSessionId should point to global session if there is one 1`] = `
[DetoxConfigError: session.sessionId property should be a non-empty string

HINT: Check that in your Detox config at path:
/home/detox/myproject/.detoxrc.json

{
  session: {
    server: 'ws://localhost:12837'
  }
}]
`;

exports[`DetoxConfigErrorBuilder .malformedAppLaunchArgs should create an error with specifying the config name 1`] = `
[Error: TODO - malformedAppLaunchArgs
{}]
`;

exports[`DetoxConfigErrorBuilder .malformedUtilBinaryPaths should create an error with specifying the config name 1`] = `
[Error: TODO - malformedUtilBinaryPaths
{}]
`;

exports[`DetoxConfigErrorBuilder .missingBuildScript should create a generic error, if the config location is not known 1`] = `
[DetoxConfigError: Could not find a build script inside "android.release" configuration.

HINT: Check contents of your Detox config.

{
  configurations: {
    'android.release': {
      type: 'android.emulator',
      device: [Object]
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .missingBuildScript should create an error with a hint, if the config location is known 1`] = `
[DetoxConfigError: Could not find a build script inside "android.release" configuration.

HINT: Check contents of your Detox config at path:
/home/detox/myproject/.detoxrc.json

{
  configurations: {
    'android.release': {
      type: 'android.emulator',
      device: [Object]
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .missingDeviceProperty should include the config location into a hint message if it is known 1`] = `
[DetoxConfigError: Missing or empty "device" property inside the device config.

HINT: It should have the device query to run on, e.g.:

{
  "type": "android.emulator",
  "device": { "foo": ... }
      // or { "bar": ... }
}
Check that in your Detox config at path:
/home/detox/myproject/.detoxrc.json

{
  configurations: {
    'android.release': {
      type: 'android.emulator'
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .missingDeviceProperty should work with aliased configurations 1`] = `
[DetoxConfigError: Missing or empty "device" property inside the device config.

HINT: It should have the device query to run on, e.g.:

{
  "type": "android.emulator",
  "device": { "foo": ... }
      // or { "bar": ... }
}
Check that in your Detox config.

{
  devices: {
    emulator: {
      type: 'android.emulator'
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .missingDeviceProperty should work with plain configurations 1`] = `
[DetoxConfigError: Missing or empty "device" property inside the device config.

HINT: It should have the device query to run on, e.g.:

{
  "type": "android.emulator",
  "device": { "foo": ... }
      // or { "bar": ... }
}
Check that in your Detox config.

{
  configurations: {
    'android.release': {
      type: 'android.emulator'
    }
  }
}]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationAtGivenPath should create an error with the attempted config path 1`] = `
[DetoxConfigError: Failed to find Detox config at:
/home/detox/myproject/.detoxrc.json

HINT: Make sure the specified path is correct.]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationSpecified should create a generic error, if the configuration file is not a package.json 1`] = `
[DetoxConfigError: Cannot run Detox without a configuration.

HINT: Make sure to create external .detoxrc.json configuration in the working directory before you run Detox.]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationSpecified should create a generic error, if the configuration file is unknown 1`] = `
[DetoxConfigError: Cannot run Detox without a configuration.

HINT: Make sure to create external .detoxrc.json configuration in the working directory before you run Detox.]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationSpecified should create an error with a package.json hint, if THAT is the configuration file 1`] = `
[DetoxConfigError: Cannot run Detox without a configuration.

HINT: Create an external .detoxrc.json configuration, or add "detox" configuration section to your package.json at:
/home/detox/myproject/package.json]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationWithGivenName should create a generic error, if the config location is not known 1`] = `
[DetoxConfigError: Failed to find a configuration named "otherConf" in Detox config.

HINT: Below are the configurations Detox was able to find:
* conf1]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationWithGivenName should create an error with a hint, if the config location is known 1`] = `
[DetoxConfigError: Failed to find a configuration named "otherConf" in Detox config at path:
/etc/detox/config.js

HINT: Below are the configurations Detox was able to find:
* conf1]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationsInside should create a generic error if all is unknown 1`] = `
[DetoxConfigError: There are no configurations in the given Detox config.

{
  configurations: undefined
}]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationsInside should create an error with Detox config fragment, if the path is not known 1`] = `
[DetoxConfigError: There are no configurations in the given Detox config.

{
  configurations: undefined,
  config: {
    ios: {},
    android: {}
  }
}]
`;

exports[`DetoxConfigErrorBuilder .noConfigurationsInside should create an error with Detox config location hint, if it is known 1`] = `
[DetoxConfigError: There are no configurations in the given Detox config.

HINT: Examine the config at path:
/home/detox/myproject/.detoxrc.json

{
  configurations: undefined
}]
`;

exports[`DetoxConfigErrorBuilder .thereAreNoDeviceConfigs should create an error with a hint: with config path 1`] = `
[DetoxConfigError: Cannot use aliases since there is no "devices" config in Detox config at path:
/etc/detox/config.js

HINT: You should create a dictionary of device configurations in Detox config, e.g.:
{
  "devices": {
*-> "simulator": {
|     "type": "ios.simulator", // or "android.emulator", or etc...
|     "device": { "type": "iPhone 12" }, // or e.g.: { "avdName": "Pixel_API_29" }
|   }
| },
| "configurations": {
|   "conf1": {
*---- "device": "simulator",
      ...
    }
  }
}
]
`;

exports[`DetoxConfigErrorBuilder .thereAreNoDeviceConfigs should create an error with a hint: without config path 1`] = `
[DetoxConfigError: Cannot use aliases since there is no "devices" config in Detox config.

HINT: You should create a dictionary of device configurations in Detox config, e.g.:
{
  "devices": {
*-> "simulator": {
|     "type": "ios.simulator", // or "android.emulator", or etc...
|     "device": { "type": "iPhone 12" }, // or e.g.: { "avdName": "Pixel_API_29" }
|   }
| },
| "configurations": {
|   "conf1": {
*---- "device": "simulator",
      ...
    }
  }
}
]
`;
